File: GrooveOn.Models/RequestObjects/AlbumUpsertRequest.cs
using System;
using System.Collections.Generic;
using GrooveOn.Model.RequestObjects;

namespace GrooveOn.Model.Requests
{
    public class AlbumUpsertRequest
    {
        public string ExternalAlbumId { get; set; } = string.Empty;
        public string? ExternalArtistId { get; set; }

        public string Source { get; set; } = "Deezer";

        public string Title { get; set; } = string.Empty;
        public string ArtistName { get; set; } = string.Empty;

        public string? CoverUrl { get; set; }
        public string? Description { get; set; }
        public DateTime? ReleaseDate { get; set; }

        public List<GenreUpsertRequest> Genres { get; set; }

        public List<SongUpsertRequest> Tracks { get; set; }
    }
}

File: UI/grooveon_desktop/lib/models/request/album_upsert_request.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:grooveon_desktop/models/request/song_upsert_request.dart';
import 'package:grooveon_desktop/models/request/genre_upsert_request.dart';

part 'album_upsert_request.g.dart';

@JsonSerializable(explicitToJson: true)
class AlbumUpsertRequest {
  final String externalAlbumId;
  final String? externalArtistId;
  final String source;
  final String title;
  final String artistName;
  final String? coverUrl;
  final String? description;
  final DateTime? releaseDate;

  final List<GenreUpsertRequest> genres; 

  final List<SongUpsertRequest> tracks;

  AlbumUpsertRequest({
    required this.externalAlbumId,
    this.externalArtistId,
    this.source = "Deezer",
    required this.title,
    required this.artistName,
    this.coverUrl,
    this.description,
    this.releaseDate,
    this.genres = const [],
    required this.tracks,
  });

  factory AlbumUpsertRequest.fromJson(Map<String, dynamic> json) =>
      _$AlbumUpsertRequestFromJson(json);

  Map<String, dynamic> toJson() => _$AlbumUpsertRequestToJson(this);
}